src/per/qspi.cpp

```diff
@@ -0,0 +1,1008 @@
+
+    if(HAL_QSPI_Init(&halqspi_) != HAL_OK)
+    {
+        return QSPIHandle::Result::ERR;
+    }
+    if(ResetMemory() != QSPIHandle::Result::OK)
+    {
+        return QSPIHandle::Result::ERR;
+    }
+    //    uint8_t fifothresh = HAL_QSPI_GetFifoThreshold(&dsy_qspi_handle.hqspi);
+    //    uint8_t reg = 0;
+    //    reg = get_status_register(&dsy_qspi_handle);
+    if(DummyCyclesConfig(device) != QSPIHandle::Result::OK)
+    {
+        return QSPIHandle::Result::ERR;
+    }
+    // Once writing test with 1 Line is confirmed lets move this out, and update writing to use 4-line.
+    if(QuadEnable() != QSPIHandle::Result::OK)
+    {
+        return QSPIHandle::Result::ERR;
+    }
+
+    if(mode_ == Mode::MEMORY_MAPPED)
+    {
+        if(EnableMemoryMappedMode() != QSPIHandle::Result::OK)
+        {
+            return QSPIHandle::Result::ERR;
+        }
+    }
+
+    return QSPIHandle::Result::OK;
+}
+
+
+QSPIHandle::Result QSPIHandle::Impl::Deinit()
+{
+    halqspi_.Instance = QUADSPI;
+    if(HAL_QSPI_DeInit(&halqspi_) != HAL_OK)
+    {
+        return QSPIHandle::Result::ERR;
+    }
+    HAL_QSPI_MspDeInit(&halqspi_);
+    return QSPIHandle::Result::OK;
+}
+
+
+QSPIHandle::Result QSPIHandle::Impl::WritePage(uint32_t address,
+                                               uint32_t size,
+                                               uint8_t* buffer,
+                                               bool     reset_mode)
+{
+    SET_MODE(Mode::INDIRECT_POLLING);
+
+    QSPI_CommandTypeDef s_command;
+    s_command.InstructionMode   = QSPI_INSTRUCTION_1_LINE;
+    s_command.Instruction       = PAGE_PROG_CMD;
+    s_command.AddressMode       = QSPI_ADDRESS_1_LINE;
+    s_command.AddressSize       = QSPI_ADDRESS_24_BITS;
+    s_command.AlternateByteMode = QSPI_ALTERNATE_BYTES_NONE;
+    s_command.DataMode          = QSPI_DATA_1_LINE;
+    s_command.DummyCycles       = 0;
+    s_command.NbData            = size <= 256 ? size : 256;
+    s_command.DdrMode           = QSPI_DDR_MODE_DISABLE;
+    s_command.DdrHoldHalfCycle  = QSPI_DDR_HHC_ANALOG_DELAY;
+    s_command.SIOOMode          = QSPI_SIOO_INST_EVERY_CMD;
+    s_command.Address           = address;
+    if(WriteEnable() != QSPIHandle::Result::OK)
+    {
+        goto write_error;
+    }
+    if(HAL_QSPI_Command(&halqspi_, &s_command, HAL_QPSI_TIMEOUT_DEFAULT_VALUE)
+       != HAL_OK)
+    {
+        goto write_error;
+    }
+    if(HAL_QSPI_Transmit(
+           &halqspi_, (uint8_t*)buffer, HAL_QPSI_TIMEOUT_DEFAULT_VALUE)
+       != HAL_OK)
+    {
+        goto write_error;
+    }
+    if(AutopollingMemReady(HAL_QPSI_TIMEOUT_DEFAULT_VALUE)
+       != QSPIHandle::Result::OK)
+    {
+        goto write_error;
+    }
+
+    if(reset_mode)
+        SET_MODE(Mode::MEMORY_MAPPED);
+    return QSPIHandle::Result::OK;
+
+// I'm sorry, but it does minimize code size here
+write_error:
+    if(reset_mode)
+        SET_MODE(Mode::MEMORY_MAPPED);
+    return QSPIHandle::Result::ERR;
```

Comment:
how much code size, because I really don't like this.

even just doing:
```
status = writeenable();
if (status == OK)
  status = command()
if (status == OK)
  status = transmit()
. . . etc.
if (reset_mode)
  SetMode(Mode::MemoryMapped);
return status;
```

looks like it be as cheap, or just barely more instructions.